### PR TITLE
ferretdb: 0.7.1 -> 0.9.0

### PR DESCRIPTION
**Pull request by AlekSi**
###### Description of changes

Tweak the build process to set correct version.

(puts on FerretDB developer hat) But I'm also a bit worried about how this package is built. `internal/util/version/gen/version.txt` was supposed to be an internal detail of our build process, not something packagers should rely on. I want to invite NixOS community members to [our community](https://github.com/FerretDB/FerretDB#community) to discuss how we could provide a stable ground for packagers in general and nixpkgs in particular.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by dit7ya — APPROVED** [review]
Thanks for this PR @AlekSi, it is of course great to have feedback from the package authors themselves. Do you think there is any other (non-internal) nicer way to include these values?

**Comment by AlekSi** [comment]
@dit7ya Sorry for the long delay. We added basic documentation for packagers [there](https://github.com/FerretDB/FerretDB#building-and-packaging) and a bit more information about files needed for build and Go build tags [there](https://pkg.go.dev/github.com/FerretDB/FerretDB/build/version).

I think it is fine to use `buildGoModule` for now, but in the future, it would be better to use our recommended approach (`task build-release`).

That PR should be updated to use newer SHA256 hashes. Also, is it possible to have a "test script" to ensure the build succeeds? Running `ferretdb --version` should be enough, I think.

**Comment by ChanceHarrison** [comment]
> but in the future, it would be better to use our recommended approach (task build-release). @AlekSi 

Better in what ways?

I tried to do just that (use tasks for building) and the primary problem that I ran into was with the [gen-version task](https://github.com/FerretDB/FerretDB/blob/main/build/version/generate.go). Specifically, there is no git repo to run git commands against when `fetchFromGithub` is used as it typically only fetches archives, so that leaves that task (and thus the build) dead in the water. 

I tried using `fetchGit` with the `deepClone` argument set to `true` (which also implies the `leaveDotGit` argument) in hopes that it would be a (suboptimal *) workaround, but `git describe --dirty` failed. Probably because a shallow clone was done despite the `deepClone` argument (see https://github.com/NixOS/nixpkgs/issues/179788). 

>  \* suboptimal because it seems that leaving the `.git` directory leads to [reproducibility issues](https://github.com/NixOS/nixpkgs/issues/8567).

Strike two, then. I'm not sure how get the task working without a workaround for `fetchGit` and `deepClone` (which I'm not hopeful for since it seems like a seldom-used and not well-liked argument). 

It seems then that the task route, while ensuring that packages are built consistently regardless of distro seems to be a bit of a (Nix-specific) quagmire.

As for what the best way forward is, I'm afraid I must defer to others with more Nix packaging experience.

In case anyone is interested in reproducing my build attempt using the package's built-in [tasks](https://github.com/FerretDB/FerretDB/blob/main/Taskfile.yml):

```nix
{ lib, buildGoModule, fetchgit, go-task, git }:

buildGoModule rec {
  pname = "ferretdb";
  version = "0.8.0";

  src = fetchgit {
    url = "https://github.com/FerretDB/FerretDB";
    rev = "v${version}";
    sha256 = "sha256-BiTVTFep7xF1OhMh/I0RzK/mDJLunaU6vfuq33Yga3I=";
    deepClone = true;
  };

  vendorSha256 = "sha256-VWDRwf3pnD+Q516d05kfm0iPUuSHs00QKGYN4rqNfw4=";

  nativeBuildInputs = [ go-task git ]; # Trying to build the task bin using `go generate tools/tools.go` did not work at all due to the way dependencies are handled/vendored.

  # override the standard buildPhase of buildGoModule
  buildPhase = ''
  task build-release
  '';

  # meta = ...
}
```
building that results in...
```
[...snip...]
building
task: [gen-version] go generate -x ./build/version
go run ./generate.go
fatal: No names found, cannot describe anything.
branch.txt: fetchgit
commit.txt: c3230915b4462373b6591ec56aa8b13eb04215a3
panic: Failed to run "git describe --dirty": exit status 128

goroutine 18 [running]:
main.runGit({0xc0000527a8?, 0x0?, 0x0?})
        ./generate.go:37 +0x11b
main.main.func1()
        ./generate.go:59 +0x9b
created by main.main
        ./generate.go:56 +0x85
exit status 2
build/version/version.go:58: running "go": exit status 1
```

File: pkgs/servers/nosql/ferretdb/default.nix
Comment (ChanceHarrison) [inline pkgs/servers/nosql/ferretdb/default.nix:8]: Changing the version means that the SRI hashes for the GitHub fetch (line 14) and the Go dependencies (line 22) need to change accordingly as well.

File: pkgs/servers/nosql/ferretdb/default.nix
Comment (ChanceHarrison) [inline pkgs/servers/nosql/ferretdb/default.nix:19]: There is no `nixpkgs` branch in the FerretDB repo, right? Am I interpreting the use of the term "branch" too literally?

From [version.go](https://github.com/FerretDB/FerretDB/blob/c3230915b4462373b6591ec56aa8b13eb04215a3/build/version/version.go#L19):

```go
// The following generated text files may be present in this (build/version) directory during building:
//   - version.txt (required) contains information about the FerretDB version in a format
//     similar to `git describe` output: `v<major>.<minor>.<patch>`.
//   - commit.txt (optional) contains information about the source git commit.
//   - branch.txt (optional) contains information about the source git branch.
```

**Review by ChanceHarrison — CHANGES_REQUESTED** [review]

**Review by dit7ya — CHANGES_REQUESTED** [review]
@AlekSi Let me know if you want me to help you with changing the hashes as mentioned in the previous review.

**Comment by noisersup** [comment]
(Another FerretDB dev here) 
We've updated sha256 for src and vendors. Doing that, we've also noticed that the current derivation doesn't include Tigris support, which is an alternative backend for PostgreSQL, so we've updated build tags to do so.

**Comment by dit7ya** [comment]
Result of `nixpkgs-review pr 202559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ferretdb</li>
  </ul>
</details>

File: pkgs/servers/nosql/ferretdb/default.nix
Comment (dit7ya) [reply pkgs/servers/nosql/ferretdb/default.nix:19]: @noisersup Could you please address this?

Comment (AlekSi) [reply pkgs/servers/nosql/ferretdb/default.nix]: I updated that PR to use a new file that will be used by some future FerretDB release; see FerretDB/FerretDB#1806.

**Comment by AlekSi** [comment]
Hm… Not sure why re-review requests are being removed.


@ChanceHarrison @dit7ya PTAL

**Comment by dit7ya** [comment]
Result of `nixpkgs-review pr 202559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ferretdb</li>
  </ul>
</details>

**Review by dit7ya — COMMENTED** [review]
I think it is okay to squash the two commits as `ferredb: 0.7.1 -> 0.8.1`. Otherwise LGTM.

**Comment by AlekSi** [comment]
@dit7ya Done, PTAL

**Comment by dit7ya** [comment]
I think the commit message should be `ferredb: 0.7.1 -> 0.8.1` as it is more importantly a version bump (than a refactor).

**Comment by dit7ya** [comment]
Result of `nixpkgs-review pr 202559` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ferretdb</li>
  </ul>
</details>

**Review by dit7ya — APPROVED** [review]
LGTM.

**Review by ChanceHarrison — APPROVED** [review]
+1 LGTM

```
[~/nixpkgs]$ nix run nixpkgs#nixpkgs-review pr 202559
...
1 package built:
ferretdb
...
[nix-shell:~/.cache/nixpkgs-review/pr-202559-1]$ ferretdb --version
version: v0.9.0
commit: unknown
branch: unknown
dirty: false
package: nixpkgs
debugBuild: false
```

**Comment by AlekSi** [comment]
What should be done to merge this PR?

**Comment by dit7ya** [comment]
Given the shortage of committers in nixpkgs, we must wait - unfortunately. You can post on this thread after (broadly) one week to draw attention  - https://discourse.nixos.org/t/prs-already-reviewed/2617.

**Review by noisersup — APPROVED** [review]

**Review by dit7ya — APPROVED** [review]

**Review by noisersup — APPROVED** [review]

**Pull request merged**
